### ISSUE TITLE
[Question] how to determine Tz's scope?

### ISSUE THREAD

**Issue by LeroyChou** [dingdingcai/SC6D-pose#2]
Hi dingdingcai,
    I've read your paper. It's such an amazing job! 
    In your job, you replace regressing tz with classfiying tz. Could you tell me how you determine the nearest(or farthest) Tz, to which you set a value [here](https://github.com/dingdingcai/SC6D-pose/blob/7c5cf9cad7c150478215a12519d3a21d1995e5e6/config.py#L48). Many thanks.
Best, 
Leroy

**Comment by dingdingcai** [comment]
Thanks for your interests to our work! As mentioned in Section 3.3 of the paper, the lower and upper bounds of Tz are calcuated from the training dataset.

**Comment by LeroyChou** [comment]
So you went through the training dataset and counted the max and min values, right?

I also read the paper DORN and I find some differences between DORN and SC6D.
1. discretization of depth interval. DORN adapts  spacing-increasing discretization while you adapt uniform discretization. 
2. loss function. DORN uses ordinal loss as while you use focal loss.

Could you tell me why do you make such changes?

**Comment by dingdingcai** [comment]
> 



> So you went through the training dataset and counted the max and min values, right?
> 
> I also read the paper DORN and I find some differences between DORN and SC6D.
> 
> 1. discretization of depth interval. DORN adapts  spacing-increasing discretization while you adapt uniform discretization.
> 2. loss function. DORN uses ordinal loss as while you use focal loss.
> 
> Could you tell me why do you make such changes?

1. I compute the min and max values of Tz based on the object-centric image crops, then discretize the Tz into uniform bins.
2. As the target objects are usually distributed within a very limited range (e.g., < 2m), not like the case in DORN (>>2m), so uniform discretization should be fine in our case.
3. I utilize the dynamic zoom-in strategy (as in  CDPN, GDR-Net) to generate the object-centric images for training, and (Tz-) class labels are not uniformly distributed. Thus, I use FocalLoss to handle this unbalanced classification task.

**Comment by LeroyChou** [comment]
> 
Thank you. I'm still confused about the 3rd point. Indeed, focal loss is a good choice in the case of unbalanced data. But it ignores the labels' ranked structure, which might be import in my guess. Why don't combine those two losses? Or my guess is wrong?

**Comment by LeroyChou** [comment]
> > 
> 
> > So you went through the training dataset and counted the max and min values, right?
> > I also read the paper DORN and I find some differences between DORN and SC6D.
> > 
> > 1. discretization of depth interval. DORN adapts  spacing-increasing discretization while you adapt uniform discretization.
> > 2. loss function. DORN uses ordinal loss as while you use focal loss.
> > 
> > Could you tell me why do you make such changes?
> 
> 1. I compute the min and max values of Tz based on the object-centric image crops, then discretize the Tz into uniform bins.
> 2. As the target objects are usually distributed within a very limited range (e.g., < 2m), not like the case in DORN (>>2m), so uniform discretization should be fine in our case.
> 3. I utilize the dynamic zoom-in strategy (as in  CDPN, GDR-Net) to generate the object-centric images for training, and (Tz-) class labels are not uniformly distributed. Thus, I use FocalLoss to handle this unbalanced classification task.

Thank you. I'm still confused about the 3rd point. Indeed, focal loss is a good choice in the case of unbalanced data. But it ignores the labels' ranked structure, which might be import in my guess. Why not combine those two losses? Or my guess is wrong?

**Comment by dingdingcai** [comment]
> 



> > > 
> > 
> > 
> > > So you went through the training dataset and counted the max and min values, right?
> > > I also read the paper DORN and I find some differences between DORN and SC6D.
> > > 
> > > 1. discretization of depth interval. DORN adapts  spacing-increasing discretization while you adapt uniform discretization.
> > > 2. loss function. DORN uses ordinal loss as while you use focal loss.
> > > 
> > > Could you tell me why do you make such changes?
> > 
> > 
> > 
> > 1. I compute the min and max values of Tz based on the object-centric image crops, then discretize the Tz into uniform bins.
> > 2. As the target objects are usually distributed within a very limited range (e.g., < 2m), not like the case in DORN (>>2m), so uniform discretization should be fine in our case.
> > 3. I utilize the dynamic zoom-in strategy (as in  CDPN, GDR-Net) to generate the object-centric images for training, and (Tz-) class labels are not uniformly distributed. Thus, I use FocalLoss to handle this unbalanced classification task.
> 
> Thank you. I'm still confused about the 3rd point. Indeed, focal loss is a good choice in the case of unbalanced data. But it ignores the labels' ranked structure, which might be import in my guess. Why not combine those two losses? Or my guess is wrong?

In DORN, the task is to estimate the dense pixel-wise depth values, and thesee depth values usually have very strong ordinal correlations among each other (these values can be ordered). But in SC6D, we aim to estimate a _single_ distance value to the object center (it does not make any sense to rank a single value).

**Comment by LeroyChou** [comment]
Sorry for the late reply. I ran two experiments, one with an ordinal loss and the other with a focal loss. It seems that ordinal loss leads to better results than focal loss. Have you done these comparative experiments? Also, none of them are better than regressing Tz directly, which is different from your conclusion.

**Comment by dingdingcai** [comment]
Thanks for your experiments!  Would you mind sharing your experimental results?  I didn't experiment with the ordinal loss, but in the ablation study, the classification-based result outperformed the regression-based one by a large margin on T-LESS (0.658 vs. 0.716).

**Comment by LeroyChou** [comment]
Of course I don't mind. I ran 100k iterations on linemod to classify delta_tz in two experiments. The delta_tz loss is 0.02395 for the ordinal loss experiment and 0.04922 for the focal loss experiment. I also did an experiment with regressing delta_tz and its loss is 0.01874. Note that I'm not basing the experiment on your code, but on our own project, which is not yet open. But I don't think that changes the conclusion.
Have you tried your ablation study on linemod?

**Comment by dingdingcai** [comment]
No, I haven't tried it on LineMOD. Note that I evaluated the performance based on the standard 6D pose metrics. However, you are comparing the delta_Tz training losses calculated from two different objective functions, which does not mean the performance of the regression-based one is better than the classification-based one.
